File: Prac/cpp/data_structure/linkedList.cpp
#include <iostream>
#include <set>
using namespace std;

struct Node{
    int data;
    Node* next;
};

class LinkedList {
    private: 
        Node* head;
        Node* tail;

    public:
        LinkedList(){
            head = nullptr;
            tail = nullptr;
        };

        void insert(int n){
            Node* newNode = new Node;
            newNode->data = n;
            newNode->next = nullptr;
            
            if(head == nullptr){
                head = newNode;
                tail = newNode;
            }else {
                tail->next = newNode;
                tail = tail->next;
            }
        }

        void print(){
            Node* ptr = head;

            while(ptr != nullptr){
                cout << ptr->data << " ";
                ptr = ptr->next;
            }

            cout << "\n";
        }
};

int main() {
    
    LinkedList ll = LinkedList();
    ll.insert(1);
    ll.insert(2);
    ll.insert(3);
    ll.insert(4);
    ll.insert(2);
    ll.print();


    return 0;
}

File: Prac/python/data_structure/linkedlist.py
class Node:
    def __init__(self, d):
        self.data = d
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None
        self.tail = None

    def add(self, val):
        node = Node(val)
        if(self.head is None):
            self.head = node
            self.tail = node
        else:
            self.tail.next = node
            self.tail = self.tail.next

    def show(self):
        ptr = self.head
        s = ""
        while (ptr != None):
            s += str(ptr.data)
            s += " "
            ptr = ptr.next
        print(s)

    def find(self, val):
        ptr = self.head
        while (ptr != None):
            if(ptr.data == val):
                return ptr
            ptr = ptr.next
        return None

    def removeNth(self, idx):
        if(idx == 1):
            self.head = self.head.next
        else:
            before = self.head
            for i in range(idx - 2):
                before = before.next
            before.next = before.next.next
            if(before.next == None):
                self.tail = before


ll = LinkedList()
for i in range(10):
    ll.add(i)
ll.show()

for i in range(10):
    ll.removeNth(1)
ll.show()

ll.add(1)
ll.add(2)

ll.show()